homeassistant/helpers/sun.py

```diff
@@ -0,0 +1,82 @@
+"""Helpers for sun events."""
+from datetime import timedelta
+
+from homeassistant.core import callback
+from homeassistant.util import dt as dt_util
+
+
+# Cache astral locations so they aren't recreated with the same args
+_LOCATION_CACHE = {}
+
+
+@callback
+def get_astral_location(hass):
+    """Get an astral location for the current hass configuration."""
+    from astral import Location
+
+    latitude = hass.config.latitude
+    longitude = hass.config.longitude
+    timezone = hass.config.time_zone.zone
+    elevation = hass.config.elevation
+    info = ('', '', latitude, longitude, timezone, elevation)
+
+    if info not in _LOCATION_CACHE:
+        _LOCATION_CACHE[info] = Location(info)
+
+    return _LOCATION_CACHE[info]
+
+
+@callback
+def get_astral_event_next(hass, event, utc_point_in_time=None, offset=None):
+    """Calculate the next specified solar event."""
+    import astral
+
+    location = get_astral_location(hass)
+
+    if offset is None:
+        offset = timedelta()
+
+    if utc_point_in_time is None:
+        utc_point_in_time = dt_util.utcnow()
+
+    mod = -1
+    while True:
+        try:
+            next_dt = getattr(location, event)(
+                dt_util.as_local(utc_point_in_time).date() +
+                timedelta(days=mod),
+                local=False) + offset
+            if next_dt > utc_point_in_time:
+                return next_dt
+        except astral.AstralError:
+            pass
+        mod += 1
+
+
+@callback
+def get_astral_event_date(hass, event, now=None):
```

Comment:
Sounds good. For now I'd like to keep a provision in this helper in case a datetime is passed. We discovered astral had the particularly nasty problem where it would accept datetimes instead of dates, and would return a result that was close, but not exactly right, since it was basically a fractional date. This way we can ensure anything hass is passing to astral is sanitized.